src/leap/mail/adaptors/soledad.py

```diff
@@ -0,0 +1,1047 @@
+        leap_assert_type(self.indexes, dict)
+
+        def _create_index(name, expression):
+            return store.create_index(name, *expression)
+
+        def init_idexes(indexes):
+            deferreds = []
+            db_indexes = dict(indexes)
+            # Loop through the indexes we expect to find.
+            for name, expression in self.indexes.items():
+                if name not in db_indexes:
+                    # The index does not yet exist.
+                    d = _create_index(name, expression)
+                    deferreds.append(d)
+                elif expression != db_indexes[name]:
+                    # The index exists but the definition is not what expected,
+                    # so we delete it and add the proper index expression.
+                    d = store.delete_index(name)
+                    d.addCallback(
+                        lambda _: _create_index(name, *expression))
+                    deferreds.append(d)
+            return defer.gatherResults(deferreds, consumeErrors=True)
+
+        def store_ready(whatever):
+            self.store_ready = True
+            return whatever
+
+        self.deferred_indexes = store.list_indexes()
+        self.deferred_indexes.addCallback(init_idexes)
+        self.deferred_indexes.addCallback(store_ready)
+        return self.deferred_indexes
+
+    def _wait_for_indexes(self):
+        """
+        Make the marked methods to wait for the indexes to be ready.
+        Heavily based on
+        http://blogs.fluidinfo.com/terry/2009/05/11/a-mixin-class-allowing-python-__init__-methods-to-work-with-twisted-deferreds/
+
+        :param methods: methods that need to wait for the indexes to be ready
+        :type methods: tuple(str)
+        """
+        leap_assert_type(self.wait_for_indexes, list)
+        methods = self.wait_for_indexes
+
+        self.waiting = []
+        self.stored = {}
+
+        def makeWrapper(method):
+            def wrapper(*args, **kw):
+                d = defer.Deferred()
+                d.addCallback(lambda _: self.stored[method](*args, **kw))
+                self.waiting.append(d)
+                return d
+            return wrapper
+
+        for method in methods:
+            self.stored[method] = getattr(self, method)
+            setattr(self, method, makeWrapper(method))
+
+    def _restore_waiting_methods(self, _):
+        for method in self.stored:
+            setattr(self, method, self.stored[method])
+        for d in self.waiting:
+            d.callback(None)
+
+
+class SoledadMailAdaptor(SoledadIndexMixin):
+
+    implements(IMailAdaptor)
+    store = None
+
+    indexes = indexes.MAIL_INDEXES
+    wait_for_indexes = ['get_or_create_mbox', 'update_mbox', 'get_all_mboxes']
+
+    mboxwrapper_klass = MailboxWrapper
+
+    def __init__(self):
+        SoledadIndexMixin.__init__(self)
+
+    mboxwrapper_klass = MailboxWrapper
+
+    def __init__(self):
+        SoledadIndexMixin.__init__(self)
```

Comment:
Two `__init__` a two `mboxwrapper_klass` with the same params and the same implementation? Is there something I'm missing or this is a copy-paste failure?